File: ruby/isbn-verifier/isbn_verifier.rb
class IsbnVerifier
  def self.valid?(str)
    str !~ /^\d-?\d{3}-?\d{5}-?[\dX]$/ ? false :
    str.scan(/\w/).reverse.each_with_index
      .map { _1 == 'X' ? 10 : _1.to_i * (_2 + 1) }.sum % 11 == 0
  end
end

File: lua/isbn-verifier/isbn-verifier.lua
return {
  valid = function(isbn)
    if not isbn:match("^%d[-]?%d%d%d[-]?%d%d%d%d%d[-]?[%dX]$") then
        return false
    end
    isbn = string.reverse(isbn:gsub("[^%dX]", ""))
    local sum = 0
    for i=1, #isbn do
      local char = isbn:sub(i,i)
      if char:match("X") then
          sum = sum + 10
      end
      if char:match("%d") then
        sum = sum + char * i
      end
    end
    return sum % 11 == 0
  end
}
